includes/gateway/class-omise-payment-billpayment-tesco.php

```diff
@@ -0,0 +1,148 @@
+		 */
+		public function init_form_fields() {
+			$this->form_fields = array(
+				'enabled' => array(
+					'title'   => __( 'Enable/Disable', 'omise' ),
+					'type'    => 'checkbox',
+					'label'   => __( 'Enable Omise Tesco Bill Payment', 'omise' ),
+					'default' => 'no'
+				),
+
+				'title' => array(
+					'title'       => __( 'Title', 'omise' ),
+					'type'        => 'text',
+					'description' => __( 'This controls the title which the user sees during checkout.', 'omise' ),
+					'default'     => __( 'Bill Payment: Tesco', 'omise' ),
+				),
+
+				'description' => array(
+					'title'       => __( 'Description', 'omise' ),
+					'type'        => 'textarea',
+					'description' => __( 'This controls the description which the user sees during checkout.', 'omise' )
+				),
+			);
+		}
+
+		/**
+		 * @inheritdoc
+		 */
+		public function charge( $order_id, $order ) {
+			$metadata = array_merge(
+				apply_filters( 'omise_charge_params_metadata', array(), $order ),
+				array( 'order_id' => $order_id ) // override order_id as a reference for webhook handlers.
+			);
+
+			return OmiseCharge::create( array(
+				'amount'      => Omise_Money::to_subunit( $order->get_total(), $order->get_order_currency() ),
+				'currency'    => $order->get_order_currency(),
+				'description' => apply_filters( 'omise_charge_params_description', 'WooCommerce Order id ' . $order_id, $order ),
+				'source'      => array( 'type' => 'bill_payment_tesco_lotus' ),
+				'return_uri'  => add_query_arg( 'order_id', $order_id, site_url() . "?wc-api=omise_billpayment_tesco_callback" ),
+				'metadata'    => $metadata
+			) );
+		}
+
+		/**
+		 * @inheritdoc
+		 */
+		public function result( $order_id, $order, $charge ) {
+			if ( 'failed' == $charge['status'] ) {
+				return $this->payment_failed( $charge['failure_message'] . ' (code: ' . $charge['failure_code'] . ')' );
+			}
+
+			if ( 'pending' == $charge['status'] ) {
+				return [
+					'result'   => 'success',
+					'redirect' => $this->get_return_url( $order )
+				];
+			}
+
+			return $this->payment_failed(
+				sprintf(
+					__( 'Please feel free to try submitting your order again, or contact our support team if you have any questions (Your temporary order id is \'%s\')', 'omise' ),
+					$order_id
+				)
+			);
+		}
+
+		/**
+		 * @param  int $order_id
+		 */
+		public function display_barcode( $order_id ) {
+			if ( ! $order = $this->load_order( $order_id ) ) {
+				return;
+			}
+
+			$charge_id = $this->get_charge_id_from_order();
+			$charge    = OmiseCharge::retrieve( $charge_id );
+
+			$amount  = $charge['amount'];
+			$barcode = $charge['source']['references']['barcode'];
+			$tax_id  = $charge['source']['references']['omise_tax_id'];
+			$ref_1   = $charge['source']['references']['reference_number_1'];
+			$ref_2   = $charge['source']['references']['reference_number_2'];
+			?>
+
+			<div class="omise omise-billpayment-tesco-details">
```

Comment:
Is this a recommended way to output HTML in WooCommerce?

Reply:
@jonrandy yup (we've discussed this topic already in some PR in the past) – It is a common way in WordPress.